internal/web/cacheverifier.go

```diff
@@ -0,0 +1,225 @@
+		"(cached resource version is '%s', but expected '%s')", e.itemName, e.cachedResourceVersion, e.nonCachedResourceVersion)
+}
+
+func newVerifier(restConfig *rest.Config, lister verifyLister) *cacheVerifier {
+	return &cacheVerifier{
+		mutex:      sync.Mutex{},
+		diffMap:    make(map[string]diffMapEntry),
+		restConfig: restConfig,
+		stopCh:     make(chan struct{}),
+		errCh:      make(chan error, 1),
+		lister:     lister,
+	}
+}
+
+func (verifier *cacheVerifier) listCachedItems(ctx context.Context) ([]metav1.Object, error) {
+	return verifier.lister(ctx, verifier.cachedClient)
+}
+
+func (verifier *cacheVerifier) listNonCachedItems(ctx context.Context) ([]metav1.Object, error) {
+	return verifier.lister(ctx, verifier.nonCachedClient)
+}
+
+type verifyLister = func(ctx context.Context, client client.PackageV1Alpha1Client) ([]metav1.Object, error)
+
+var clusterPackageVerifyLister = func(ctx context.Context, client client.PackageV1Alpha1Client) ([]metav1.Object, error) {
+	var ls v1alpha1.ClusterPackageList
+	err := client.ClusterPackages().GetAll(ctx, &ls)
+	target := make([]metav1.Object, len(ls.Items))
+	for idx, item := range ls.Items {
+		target[idx] = item.GetObjectMeta()
+	}
+	return target, err
+}
+
+var packageInfoVerifyLister = func(ctx context.Context, client client.PackageV1Alpha1Client) ([]metav1.Object, error) {
+	var ls v1alpha1.PackageInfoList
+	err := client.PackageInfos().GetAll(ctx, &ls)
+	target := make([]metav1.Object, len(ls.Items))
+	for idx, item := range ls.Items {
+		target[idx] = item.GetObjectMeta()
+	}
+	return target, err
+}
+
+var packageRepoVerifyLister = func(ctx context.Context, client client.PackageV1Alpha1Client) ([]metav1.Object, error) {
+	var ls v1alpha1.PackageRepositoryList
+	err := client.PackageRepositories().GetAll(ctx, &ls)
+	target := make([]metav1.Object, len(ls.Items))
+	for idx, item := range ls.Items {
+		target[idx] = item.GetObjectMeta()
+	}
+	return target, err
+}
+
+func (verifier *cacheVerifier) setClients(cachedClient client.PackageV1Alpha1Client, nonCachedClient client.PackageV1Alpha1Client) {
+	verifier.cachedClient = cachedClient
+	verifier.nonCachedClient = nonCachedClient
+}
+
+func (verifier *cacheVerifier) start(ctx context.Context, cachedClient client.PackageV1Alpha1Client) {
+	// setup client without cache
+	nonCachedClient := client.NewOrDie(verifier.restConfig)
+	verifier.setClients(cachedClient, nonCachedClient)
+
+	tick := time.NewTicker(time.Second * time.Duration(8*(rand.Float32()+1))) // between 8 - 16 seconds
+	tickC := tick.C
+	defer tick.Stop()
+
+	var diffErr *cacheVerificationError
+
+endlessLoop:
+	for range tickC {
+		if nonCachedItems, err := verifier.listNonCachedItems(ctx); err != nil {
+			fmt.Fprintf(os.Stderr, "CACHEVERIFIER: failed to get actual items: %v\n", err)
+		} else if cachedItems, err := verifier.listCachedItems(ctx); err != nil {
+			fmt.Fprintf(os.Stderr, "CACHEVERIFIER: failed get cached items: %v\n", err)
+		} else {
```

Comment:
by using `verifier.lister(ctx, cachedClient)` and `verifier.lister(ctx, nonCachedClient)` here, we could avoid storing the clients in the struct state at all. What do you think about that?